File: worker/hub.go
//go:build js && wasm

package main

import (
	"encoding/json"
	"syscall/js"
)

// peer represents a connected WebSocket peer in the signaling hub.
type peer struct {
	wsId      int
	peerID    string
	publicKey string
	address   string
	routes    []string
}

// peers tracks all connected peers by their WebSocket ID.
var peers = make(map[int]*peer)

// peerByID maps peer IDs to WebSocket IDs for fast lookups.
var peerByID = make(map[string]int)

// peerInfo is the JSON representation of a peer in the peers list.
type peerInfo struct {
	PeerID    string   `json:"peerId"`
	PublicKey string   `json:"publicKey"`
	Address   string   `json:"address,omitempty"`
	Routes    []string `json:"routes,omitempty"`
}

// send sends a JSON message to a specific WebSocket by ID.
func send(wsId int, data []byte) {
	sendFn.Invoke(wsId, string(data))
}

// broadcast sends a JSON message to all connected peers except the sender.
func broadcast(senderWsId int, data []byte) {
	msg := string(data)
	for _, p := range peers {
		if p.wsId == senderWsId {
			continue
		}
		sendFn.Invoke(p.wsId, msg)
	}
}

// jsOnRehydrate is called by JS to silently restore a peer after hibernation.
// Unlike jsOnJoin, it does NOT send a peers list or broadcast a join notification.
// Arguments: wsId (int), peerId (string), publicKey (string), address (string), routesJSON (string)
func jsOnRehydrate(_ js.Value, args []js.Value) any {
	wsId := args[0].Int()
	peerID := args[1].String()
	publicKey := args[2].String()
	address := args[3].String()
	routes := parseRoutesJSON(args[4].String())

	peers[wsId] = &peer{
		wsId:      wsId,
		peerID:    peerID,
		publicKey: publicKey,
		address:   address,
		routes:    routes,
	}
	peerByID[peerID] = wsId

	return nil
}

// jsOnJoin is called by JS when a new peer sends a join message.
// Arguments: wsId (int), peerId (string), publicKey (string), address (string), routesJSON (string)
func jsOnJoin(_ js.Value, args []js.Value) any {
	wsId := args[0].Int()
	peerID := args[1].String()
	publicKey := args[2].String()
	address := args[3].String()
	routes := parseRoutesJSON(args[4].String())

	p := &peer{
		wsId:      wsId,
		peerID:    peerID,
		publicKey: publicKey,
		address:   address,
		routes:    routes,
	}

	// Build the current peers list before adding the new peer.
	peerInfos := make([]peerInfo, 0, len(peers))
	for _, existing := range peers {
		peerInfos = append(peerInfos, peerInfo{
			PeerID:    existing.peerID,
			PublicKey: existing.publicKey,
			Address:   existing.address,
			Routes:    existing.routes,
		})
	}

	// Add the new peer.
	peers[wsId] = p
	peerByID[peerID] = wsId

	// Send the current peers list to the new peer.
	peersMsg, _ := json.Marshal(map[string]any{
		"type":  "peers",
		"peers": peerInfos,
	})
	send(wsId, peersMsg)

	// Notify existing peers about the new arrival.
	newPeerMsg, _ := json.Marshal(map[string]any{
		"type": "peers",
		"peers": []peerInfo{{
			PeerID:    peerID,
			PublicKey: publicKey,
			Address:   address,
			Routes:    routes,
		}},
	})
	broadcast(wsId, newPeerMsg)

	return nil
}

// jsOnMessage is called by JS when a peer sends a signaling message.
// Arguments: wsId (int), rawJSON (string)
func jsOnMessage(_ js.Value, args []js.Value) any {
	rawJSON := args[1].String()

	// Parse just the routing envelope to determine the target.
	var env struct {
		Type string `json:"type"`
		To   string `json:"to"`
	}
	if err := json.Unmarshal([]byte(rawJSON), &env); err != nil {
		return nil
	}

	switch env.Type {
	case "offer", "answer", "ice-candidate":
		targetWsId, ok := peerByID[env.To]
		if ok {
			send(targetWsId, []byte(rawJSON))
		}
	}

	return nil
}

// parseRoutesJSON decodes a JSON array of route strings. Returns nil if the
// input is empty or invalid.
func parseRoutesJSON(s string) []string {
	if s == "" || s == "[]" {
		return nil
	}
	var routes []string
	if err := json.Unmarshal([]byte(s), &routes); err != nil {
		return nil
	}
	return routes
}

// jsOnLeave is called by JS when a peer disconnects.
// Arguments: wsId (int)
func jsOnLeave(_ js.Value, args []js.Value) any {
	wsId := args[0].Int()

	p, ok := peers[wsId]
	if !ok {
		return nil
	}

	peerID := p.peerID
	delete(peers, wsId)
	delete(peerByID, peerID)

	// Notify remaining peers about the departure.
	leftMsg, _ := json.Marshal(map[string]any{
		"type":   "peer-left",
		"peerId": peerID,
	})
	broadcast(-1, leftMsg)

	return nil
}

File: worker/src/hub.ts
import type { PeerInfo, SignalingMessage } from "./types.js";

interface Peer {
  wsId: number;
  peerId: string;
  publicKey: string;
  address: string;
  routes: string[];
  metadata: Record<string, string>;
}

export interface SendFn {
  (wsId: number, data: string): void;
}

export class Hub {
  private peers: Map<number, Peer> = new Map();
  private peerById: Map<string, number> = new Map();
  private sendFn: SendFn;

  constructor(sendFn: SendFn) {
    this.sendFn = sendFn;
  }

  rehydrate(
    wsId: number,
    peerId: string,
    publicKey: string,
    address: string,
    routes: string[],
    metadata: Record<string, string>
  ): void {
    const peer: Peer = {
      wsId,
      peerId,
      publicKey,
      address,
      routes,
      metadata,
    };
    this.peers.set(wsId, peer);
    this.peerById.set(peerId, wsId);
  }

  join(
    wsId: number,
    peerId: string,
    publicKey: string,
    address: string,
    routes: string[],
    metadata: Record<string, string>
  ): PeerInfo[] {
    const peer: Peer = {
      wsId,
      peerId,
      publicKey,
      address,
      routes,
      metadata,
    };

    const peerInfos: PeerInfo[] = [];
    for (const existing of this.peers.values()) {
      peerInfos.push({
        peerId: existing.peerId,
        publicKey: existing.publicKey,
        address: existing.address,
        routes: existing.routes,
        metadata: existing.metadata,
      });
    }

    this.peers.set(wsId, peer);
    this.peerById.set(peerId, wsId);

    return peerInfos;
  }

  message(_wsId: number, rawJson: string): void {
    let msg: SignalingMessage;
    try {
      msg = JSON.parse(rawJson);
    } catch {
      return;
    }

    const type = msg.type;
    const to = "to" in msg ? msg.to : null;

    if (!to) return;

    if (type === "offer" || type === "answer" || type === "ice-candidate") {
      const targetWsId = this.peerById.get(to);
      if (targetWsId !== undefined) {
        this.sendFn(targetWsId, rawJson);
      }
    }
  }

  leave(wsId: number): string | null {
    const peer = this.peers.get(wsId);
    if (!peer) return null;

    const peerId = peer.peerId;
    this.peers.delete(wsId);
    this.peerById.delete(peerId);

    return peerId;
  }

  broadcast(senderWsId: number, data: string): void {
    for (const [id] of this.peers) {
      if (id === senderWsId) continue;
      this.sendFn(id, data);
    }
  }

  getPeers(): PeerInfo[] {
    const peerInfos: PeerInfo[] = [];
    for (const peer of this.peers.values()) {
      peerInfos.push({
        peerId: peer.peerId,
        publicKey: peer.publicKey,
        address: peer.address,
        routes: peer.routes,
        metadata: peer.metadata,
      });
    }
    return peerInfos;
  }

  getPeer(wsId: number): Peer | undefined {
    return this.peers.get(wsId);
  }
}

export function createHub(sendFn: SendFn): Hub {
  return new Hub(sendFn);
}
